fedimint-client/src/db/event_log.rs

```diff
@@ -0,0 +1,499 @@
+        // TODO: This is innefficent for large logs, and there's no good operations to
+        // help? Do we just query one by one instead?
+        self.find_by_range(pos..pos.saturating_add(limit))
+            .await
+            .map(|(k, v)| {
+                (
+                    k,
+                    (
+                        v.kind,
+                        v.module,
+                        v.ts_msecs,
+                        serde_json::from_slice(&v.payload).unwrap_or_default(),
+                    ),
+                )
+            })
+            .collect()
+            .await
+    }
+}
+
+pub async fn run_event_log_ordering_task(
+    db: Database,
+    mut log_ordering_task_wakeup: watch::Receiver<()>,
+    log_event_added: watch::Sender<()>,
+) {
+    debug!(target: LOG_CLIENT_EVENT_LOG, "Event log ordering task started");
+    let mut next_entry_id = db
+        .begin_transaction_nc()
+        .await
+        .get_next_event_log_id()
+        .await;
+
+    loop {
+        let mut dbtx = db.begin_transaction().await;
+
+        let unordered_events = dbtx
+            .find_by_prefix(&UnorderedEventLogIdPrefixAll)
+            .await
+            .collect::<Vec<_>>()
+            .await;
+        trace!(target: LOG_CLIENT_EVENT_LOG, num=unordered_events.len(), "Fetched unordered events");
+
+        for (unordered_id, entry) in &unordered_events {
+            assert!(
+                dbtx.remove_entry(unordered_id).await.is_some(),
+                "Must never fail to remove entry"
+            );
+            assert!(
+                dbtx.insert_entry(&next_entry_id, entry).await.is_none(),
+                "Must never overwrite existing event"
+            );
+
+            trace!(target: LOG_CLIENT_EVENT_LOG, ?unordered_id, id=?next_entry_id, "Ordered event log event");
+            next_entry_id = next_entry_id.next();
+        }
+
+        // This thread is the only thread deleting already existing element of unordered
+        // log, And inserting new elements into ordered log, so it should never
+        // fail to commit.
+        dbtx.commit_tx().await;
+        if !unordered_events.is_empty() {
+            let _ = log_event_added.send(());
+        }
+
+        trace!(target: LOG_CLIENT_EVENT_LOG, "Event log ordering task waits for more events");
+        if log_ordering_task_wakeup.changed().await.is_err() {
+            break;
+        }
+    }
+
+    debug!(target: LOG_CLIENT_EVENT_LOG, "Event log ordering task finished");
+}
+
+pub async fn handle_events<F, R, K, V>(
+    db: Database,
+    pos_key: &K,
+    mut log_event_added: watch::Receiver<()>,
+    call_fn: F,
+) -> anyhow::Result<()>
+where
+    K: DatabaseKey + DatabaseRecord + MaybeSend + MaybeSync,
+    K: DatabaseRecord<Value = V>,
+    // Note: Into and From are a workaround for https://github.com/rust-lang/rust/issues/20041
+    V: Into<EventLogId> + From<EventLogId> + Default + MaybeSend + MaybeSync,
```

Comment:
Rust doesn't implement equality for associated types. See the comment line above. Unless I missed something.

Reply:
I don't know what that issue opener is doing, but I ran into this problem a few times and never needed to enforce equality as a separate `where` clause. Instead of doing:

```
A: Foo,
<A as Foo>::Bar = u64
```

I was always able to do this instead:

```
A: Foo<Bar = u64>
```

But I wrote very similar code to what you did here often enough before noticing that it wasn't necessary :laughing: 

The patch that shows it for this case: https://github.com/elsirion/fedimint/commit/4d444210d3f8f19314d3ca24e564fe387b203bd7

Reply:
Oh dear, I've been bamboozled. Thanks!